### PR TITLE
fix returning only visible elements

### PR DESCRIPTION
**Pull request by KazuCocoa**
ref: find elements except for name attributes in text/s, button/s #462

The problem current `text/s`, `button/s`, `textfield/s` return **ALL** elements **include hidden** elements. Previous implementation, UIAutomation based, return only **visible** elements, **Not ALL** elements.

In this PR, each methods start returning **visible(displayed)** elements. In addition, some methods also return visible elements same as previous behaviours.

Running tests on https://travis-ci.org/KazuCocoa/ruby_lib/builds/191011898

### CODE REVIEW

File: ios_tests/lib/ios/specs/ios/element/button.rb [hunk 1]
```diff
@@ -29,7 +29,6 @@ def gray
 
   t 'buttons' do
     exp = ['Back', 'Back', 'Gray', 'Right pointing arrow']
-    exp.concat ['Add contact'] if UI::Inventory.xcuitest?
```
Comment (KazuCocoa) [inline ios_tests/lib/ios/specs/ios/element/button.rb]: This element is hidden.

File: ios_tests/lib/ios/specs/ios/element/button.rb [hunk 2]
```diff
@@ -41,7 +40,7 @@ def gray
   end
 
   t 'last_button' do
-    expected = UI::Inventory.xcuitest? ? 'Add contact' : 'Rounded'
```
Comment (KazuCocoa) [inline ios_tests/lib/ios/specs/ios/element/button.rb]: "Add contact" is hidden

File: ios_tests/lib/ios/specs/ios/element/text.rb [hunk 3]
```diff
@@ -37,7 +37,7 @@ def before_first
 
   t 'texts' do
     exp = ['Controls', 'Various uses of UIControl', 'Various uses of UISegmentedControl']
-    texts.length.must_equal(UI::Inventory.xcuitest? ? 25 : 24)
```
Comment (KazuCocoa) [inline ios_tests/lib/ios/specs/ios/element/text.rb]: `25` include hidden elements.

File: lib/appium_lib/ios/element/button.rb [hunk 4]
```diff
@@ -32,8 +32,8 @@ def buttons(value = false)
       return tags button_class unless value
 
       if automation_name_is_xcuitest?
-        elements = tags button_class
-        _elements_include elements, value
+        visible_elements = tags button_class
```
Comment (KazuCocoa) [inline lib/appium_lib/ios/element/button.rb]: `tags` return only visible elements for now against XCUITest.

File: lib/appium_lib/ios/element/generic.rb [hunk 5]
```diff
@@ -38,33 +39,52 @@ def find_exact(value)
...
-    def _elements_include(elements, value)
+    # Return all elements include not displayed elements.
+    def elements_include(elements, value)
       return [] if elements.empty?
       elements.select do |element|
+        # `text` is equal to `value` if value is not `nil`
+        # `text` is equal to `name` if value is `nil`
         name = element.name
-        name.nil? ? false : name.downcase.include?(value.downcase)
+        text = element.value
```
Comment (KazuCocoa) [inline lib/appium_lib/ios/element/generic.rb]: check name and value(text) attributes.

File: lib/appium_lib/ios/element/textfield.rb [hunk 6]
```diff
@@ -27,13 +27,14 @@ def _textfields
...
-      xpath "//#{_textfields}"
+      raise_error_if_no_element _textfields_with_xpath.first
     end
 
     # @private
     # for XCUITest
     def _textfields_with_xpath
-      xpaths "//#{_textfields}"
+      elements = xpaths "//#{_textfields}"
+      select_visible_elements elements
```
Comment (KazuCocoa) [inline lib/appium_lib/ios/element/textfield.rb]: return only visible elements

File: lib/appium_lib/ios/helper.rb [hunk 7]
```diff
@@ -295,33 +295,29 @@ def find_eles_by_attr_include(class_name, attr, value)
     # @param class_name [String] the tag to match
     # @return [Element]
     def first_ele(class_name)
-      if automation_name_is_xcuitest?
-        @driver.find_element :class, class_name
-      else
-        ele_index class_name, 1
-      end
+      ele_index class_name, 1
```
Comment (KazuCocoa) [inline lib/appium_lib/ios/helper.rb]: `ele_index class_name, 1` is equal to `@driver.find_element :class, class_name`

File: lib/appium_lib/ios/helper.rb [hunk 8]
```diff
@@ -333,7 +329,8 @@ def tag(class_name)
     # @return [Element]
     def tags(class_name)
       if automation_name_is_xcuitest?
-        @driver.find_elements :class, class_name
+        elements = @driver.find_elements :class, class_name
+        select_visible_elements elements
```
Comment (KazuCocoa) [inline lib/appium_lib/ios/helper.rb]: return only visible elements

**Comment by KazuCocoa** [comment]
according to the Travis, https://travis-ci.org/KazuCocoa/ruby_lib/builds/191144663, related tests work fine.

**Comment by slipy12** [comment]
I'll check it as soon as I can!

**Comment by slipy12** [comment]
You deleted your local branch, how can I try it? I was using the gem linking to your git and pointing to branch `return_visible_elements`.

**Comment by KazuCocoa** [comment]
I have't deleted this PR's branch. 
https://github.com/KazuCocoa/ruby_lib/tree/return_visible_elements is. Could you try again? @slipy12

**Comment by slipy12** [comment]
Works as expected (tested on ios, later will test on android)

**Comment by KazuCocoa** [comment]
https://github.com/appium/ruby_lib/blob/42df79b4cd0b5a36f52628f0acdeb688c5170c0c/lib/appium_lib/ios/helper.rb#L232
failed to find elements text which have `'`. I have to parse it.

**Comment by KazuCocoa** [comment]
fixed in 2e2118f

**Comment by KazuCocoa** [comment]
I added test code for `'` in 2e2118f

**Comment by KazuCocoa** [comment]
I merge this due to all issues were fiexed.

**Comment by KazuCocoa** [comment]
@slipy12 I merge this PR. So, please use master branch if you run tests with upstream.

**Pull request merged**